CMakeLists.txt

```diff
@@ -1,85 +1,106 @@
 cmake_minimum_required(VERSION 3.3)
 
-project(rabit VERSION 0.3.0)
-
-include(CheckCXXCompilerFlag)
-CHECK_CXX_COMPILER_FLAG("-std=c++11" COMPILER_SUPPORTS_CXX11)
-if(COMPILER_SUPPORTS_CXX11)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
-else()
-  message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++11 support. Please use a different C++ compiler.")
-endif()
+project(rabit VERSION 0.3.0 LANGUAGES CXX)
 
 option(RABIT_BUILD_TESTS "Build rabit tests" OFF)
 option(RABIT_BUILD_MPI "Build MPI" OFF)
 option(RABIT_BUILD_DMLC "Include DMLC_CORE in build" OFF)
+option(RABIT_WITH_R_LIB "Fit the strict environment of R" OFF)
+
+option(DMLC_ROOT "Specify root of external dmlc core.")
+# by default point to xgboost/dmlc-core
+set(DMLC_ROOT ${CMAKE_CURRENT_LIST_DIR}/../dmlc-core)
 
 # moved from xgboost build
 if(R_LIB OR MINGW OR WIN32)
-  add_library(rabit_base src/allreduce_base.cc src/engine_base.cc src/c_api.cc)
+  add_library(rabit src/engine_empty.cc src/c_api.cc)
   set(rabit_libs rabit)
-  set_target_properties(rabit PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)
-ELSE()
+  set_target_properties(rabit
+          PROPERTIES CXX_STANDARD 11
+          CXX_STANDARD_REQUIRED ON
+          POSITION_INDEPENDENT_CODE ON)
+else()
   add_library(rabit src/allreduce_base.cc src/allreduce_robust.cc src/engine.cc src/c_api.cc)
   add_library(rabit_base src/allreduce_base.cc src/engine_base.cc src/c_api.cc)
   add_library(rabit_empty src/engine_empty.cc src/c_api.cc)
   add_library(rabit_mock_static src/allreduce_base.cc src/allreduce_robust.cc src/engine_mock.cc src/c_api.cc)
   add_library(rabit_mock SHARED src/allreduce_base.cc src/allreduce_robust.cc src/engine_mock.cc src/c_api.cc)
 
   set(rabit_libs rabit rabit_base rabit_empty rabit_mock rabit_mock_static)
-  set_target_properties(rabit rabit_base rabit_empty rabit_mock rabit_mock_static PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)
+  set_target_properties(rabit rabit_base rabit_empty rabit_mock rabit_mock_static
+    PROPERTIES CXX_STANDARD 11
+    CXX_STANDARD_REQUIRED ON
+    POSITION_INDEPENDENT_CODE ON)
 ENDIF(R_LIB OR MINGW OR WIN32)
 
 if(RABIT_BUILD_MPI)
   find_package(MPI REQUIRED)
+  if (NOT MPI_CXX_FOUND)
+    message(FATAL_ERROR "CXX Interface for MPI is required for building MPI backend.")
```

Comment:
I wrote this line, but do we actually want to support MPI backend?  I thought it's a proof of concept for the paper.  And now that the MPI CXX interface is deprecated, there will be some more works for bring it up. @chenqin What do you think?

Reply:
We might discuss this in another thread how to support MPI runtime. The rationale is there are lots of similar frameworks highly optimized such as facebook/gloo that just do MPI optimization. Not of which offers failed single worker recover. But I don't have clear picture who is using rabit other than distributed XGBoost.